File: jdroid-android-core/src/main/java/com/jdroid/android/recycler/IdentifiableDiffCallback.java
package com.jdroid.android.recycler;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.DiffUtil;


public abstract class IdentifiableDiffCallback<T> extends DiffUtil.Callback {

	private List<T> oldItems;
	private List<T> newItems;

	public IdentifiableDiffCallback(RecyclerViewAdapter recyclerViewAdapter, List<T> newItems) {
		this.oldItems = com.jdroid.java.collections.Lists.newArrayList((List<T>)recyclerViewAdapter.getItems());
		this.newItems = com.jdroid.java.collections.Lists.newArrayList(newItems);
	}

	@Override
	public int getOldListSize() {
		return oldItems.size();
	}

	@Override
	public int getNewListSize() {
		return newItems.size();
	}

	@Override
	public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
		return getId(oldItems.get(oldItemPosition)).equals(getId(newItems.get(newItemPosition)));
	}

	@Override
	public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
		return oldItems.get(oldItemPosition).equals(newItems.get(newItemPosition));
	}

	@NonNull
	protected abstract String getId(T item);
}

File: jdroid-android-core/src/main/java/com/jdroid/android/recycler/IdentifiableDiffCallback.kt
package com.jdroid.android.recycler

import androidx.recyclerview.widget.DiffUtil
import com.jdroid.java.collections.Lists

abstract class IdentifiableDiffCallback<T>(recyclerViewAdapter: RecyclerViewAdapter, newItems: List<T>) : DiffUtil.Callback() {

    private val oldItems: List<T>
    private val newItems: List<T>

    init {
        this.oldItems = Lists.newArrayList(recyclerViewAdapter.items as List<T>)
        this.newItems = Lists.newArrayList(newItems)
    }

    override fun getOldListSize(): Int {
        return oldItems.size
    }

    override fun getNewListSize(): Int {
        return newItems.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return getId(oldItems[oldItemPosition]) == getId(newItems[newItemPosition])
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldItems[oldItemPosition] == newItems[newItemPosition]
    }

    protected abstract fun getId(item: T): String
}
